fuel-vm/src/interpreter/memory.rs

```diff
@@ -19,13 +19,231 @@ use fuel_types::{
     Word,
 };
 
-use alloc::boxed::Box;
 use core::{
     ops,
-    ops::Range,
+    ops::{
+        Index,
+        IndexMut,
+        Range,
+        RangeFrom,
+        RangeTo,
+    },
 };
 
-pub type Memory<const SIZE: usize> = Box<[u8; SIZE]>;
+#[cfg(feature = "alloc")]
+use alloc::{
+    vec,
+    vec::Vec,
+};
+
+trait Heap {
+    /// Resize the heap to at least `new_len` bytes, filling the new space with `value`.
+    /// If `new_len` is less than the current length, the function does nothing.
+    /// The functions may grow the size more than `new_len` to avoid frequent
+    /// reallocations.
+    fn reverse_resize_at_least(&mut self, new_len: usize, value: u8);
+}
+
+impl Heap for Vec<u8> {
+    fn reverse_resize_at_least(&mut self, new_len: usize, value: u8) {
+        let len = self.len();
+
+        if new_len > len {
+            // It is how `Vec::resize` calculates a new capacity.
+            let cap = core::cmp::max(len * 2, new_len);
+            let cap = core::cmp::min(cap, MEM_SIZE);
+            let diff = cap - len;
+            let mut new_vec = vec![value; cap];
+            new_vec[diff..].copy_from_slice(self.as_slice());
+            *self = new_vec;
```

Comment:
This at least just copies the bytes just once, but it also forces a new allocation every time. I guess that cannot be avoided if we want to use a single `Vec`. We could maybe be a bit smarter than just doubling the size every time, e.g. starting with a kilobyte-sized allocation and not allowing the last doubling to `MEM_SIZE`, instead allocating with `reserve_exact`.

In any case, `ALOC` is currently constant cost and will allow copying the amount of entire memory, in parts. What used to be a cheap subtract and compare operation, is now a memcpy.